lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,388 @@
+    // another importer can handle.
+    if (packageName.startsWith('.') ||
+        packageName.contains('\\') ||
+        packageName.contains('%') ||
+        (packageName.startsWith('@') &&
+            !packageName.contains(p.url.separator))) {
+      return null;
+    }
+
+    var packageRoot = _resolvePackageRoot(packageName, basePath);
+
+    if (packageRoot == null) return null;
+    var jsonPath = p.join(packageRoot, 'package.json');
+
+    var jsonString = readFile(jsonPath);
+    Map<String, dynamic> packageManifest;
+    try {
+      packageManifest = json.decode(jsonString) as Map<String, dynamic>;
+    } catch (e) {
+      throw "Failed to parse $jsonPath for \"pkg:$packageName\": $e";
+    }
+
+    if (_resolvePackageExports(
+            packageRoot, subpath, packageManifest, packageName)
+        case var resolved?) {
+      if (_validExtensions.contains(p.extension(resolved))) {
+        return p.toUri(p.canonicalize(resolved));
+      } else {
+        throw "The export for '${subpath ?? "root"}' in "
+            "'$packageName' resolved to '${resolved.toString()}', "
+            "which is not a '.scss', '.sass', or '.css' file.";
+      }
+    }
+    // If no subpath, attempt to resolve `sass` or `style` key in package.json,
+    // then `index` file at package root, resolved for file extensions and
+    // partials.
+    if (subpath == null) {
+      var rootPath = _resolvePackageRootValues(packageRoot, packageManifest);
+      return rootPath != null ? p.toUri(p.canonicalize(rootPath)) : null;
+    }
+
+    // If there is a subpath, attempt to resolve the path relative to the
+    // package root, and resolve for file extensions and partials.
+    var subpathInRoot = p.join(packageRoot, subpath);
+    return FilesystemImporter.cwd.canonicalize(p.toUri(subpathInRoot));
+  }
+
+  @override
+  ImporterResult? load(Uri url) => FilesystemImporter.cwd.load(url);
+
+  /// Splits a [bare import
+  /// specifier](https://nodejs.org/api/esm.html#import-specifiers) `specifier`
+  /// into its package name and subpath, if one exists.
+  ///
+  /// Because this is a bare import specifier and not a path, we always use `/`
+  /// to avoid invalid values on non-Posix machines.
+  (String, String?) _packageNameAndSubpath(String specifier) {
+    var parts = p.url.split(specifier);
+    var name = p.fromUri(parts.removeAt(0));
+
+    if (name.startsWith('@')) {
+      if (parts.isNotEmpty) name = p.url.join(name, parts.removeAt(0));
+    }
+    var subpath = parts.isNotEmpty ? p.fromUri(p.url.joinAll(parts)) : null;
+    return (name, subpath);
+  }
+
+  /// Returns an absolute path to the root directory for the most proximate
+  /// installed `packageName`.
+  ///
+  /// Implementation of `PACKAGE_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _resolvePackageRoot(String packageName, String basePath) {
+    var baseDirectory = basePath;
```

Comment:
@nex3 `basePath` here can now either be a file (from `containingUrl` or directory (from `entryPointDirectory`). Because of this, I moved the `p.dirname` to the end, otherwise the last segment in the provided directory would never be tested. 

The downside to this is that for paths, it will now test something like `file://containingUrl.scss/node_modules/packagename`. It will never succeed, because `containingUrl.scss` can not be simultaneously a path and a directory.

I considered adding a `dirExists` check initially, but realized that would always add an additional disk read, where this only adds an additional disk read for canonicalizations with a containingUrl.

Reply:
I don't think it makes sense to handle this as either a file _or_ a directory since, as you point out, there's not a reliable way to distinguish the two without checking the filesystem. Instead, when you assign `basePath` in `canonicalize()`, add a `p.dirname()` there if it's coming from `containingUrl` so that the same variable always represents the same type of data.

Reply:
Good point- I adjusted in https://github.com/sass/dart-sass/pull/2130/commits/53472fdb880494924f06d5b0974cfac94b36b433